torch/utils/_python_dispatch.py

```diff
@@ -166,7 +166,10 @@ def transform_subclass(t, callback):
     transformed_tensors_dict = {}
     for attr in attrs:
         transformed_tensors_dict[attr] = callback(attr, getattr(t, attr))
-    return type(t).__tensor_unflatten__(transformed_tensors_dict, ctx)
+    sub = type(t).__tensor_unflatten__(transformed_tensors_dict, ctx, outer_size)
+    # TODO: nice error message for subclass implementors
+    assert sub.shape == outer_size
```

Comment:
nit: Do `definitely_true(sym_eq(sub, outer_size))` instead (it doesn't matter here because I don't think you can't actually specialize, but it's good form)

Reply:
I'm not having luck with `assert definitely_true(sym_eq(sub, outer_size))`; it breaks a test I have that torch.compiles without `dynamic=True` + passes two different completely different NT shapes.

```
AssertionError: s2 (could be from ["L['x'].size()[2]"]) not in {s1: ["L['x'].size()[1]"], s4: ["L['x'].size()[2]", "L['x']._values.size()[1]"], s3: ["L['x']._values.size()[0]"], s5: ["L['x']._offsets.size()[0]"], s0: []}.  If this assert is failing, it could be due to the issue described in https://github.com/pytorch/pytorch/pull/90665
```

what's the difference between this and `assert sub.shape == outer_size` for a layman?

Reply:
from VC: This means `__tensor_unflatten__` generated a bunch of unnecessary new symbols, instead of correctly using symbols from `old_size`.

Reply:
also from VC: it's subclass specific how the outer symbols relate to the inner ones. Seems okay to allocate new symbols for both outer and inner, then simplify via the assert here.

In the details: we want to use `==` rather than `sym_eq`. Doing the entire shape comparison symbolically ends up being too difficult to guard on and it fails..